### PR TITLE
Sampling on the flux space of multiple metabolic networks

### PR DESCRIPTION
**Pull request by hariszaf**
Two main modules are included on this PR: 
- about 6,000 bacterial metabolic models have been converted to `.mat` format using the [embl_model_convert.m](https://github.com/hariszaf/dingo/blob/pairs/embl_model_convert.m) script. The The script has been used to convert EMBL models that are publicly available [here](https://github.com/cdanielmachado/embl_gems) as well as models that were built using publicly available genomes and the [`carveme`](https://carveme.readthedocs.io/en/latest/#)  software tool for automatic metabolic modelling reconstruction 
- 2 classes called [`CommunityMetabolicNetwork`](https://github.com/hariszaf/dingo/blob/pairs/dingo/MetabolicNetwork.py#L228) and [`CommunityPolytopeSampler`](https://github.com/hariszaf/dingo/blob/pairs/dingo/PolytopeSampler.py#L302) that enable sampling on the flux space of multiple metabolic networks. 

To enable testing this new module of `dingo` an input example has been added under the [`ext_data`]( https://github.com/hariszaf/dingo/tree/pairs/ext_data/community_models) directory. 

The new module has been integrated in the `__init__.py` and can be used both from terminal and as a function of the `dingo` package.

### CODE REVIEW

File: ext_data/community_models/e_coli_core.json [hunk 1]
```diff
@@ -0,0 +1,14786 @@
+{
```
Comment (vissarion) [inline ext_data/community_models/e_coli_core.json]: is this file needed to be included in the repo? It is huge and readily available from http://bigg.ucsd.edu/static/models/e_coli_core.json

Comment (vissarion) [reply ext_data/community_models/e_coli_core.json]: similarly for `ext_data/community_models/iIT341.json` it is that large that github cannot show it

Comment (hariszaf) [reply ext_data/community_models/e_coli_core.json]: I removed the 2 `.json` files with a `README` explaining the reasond for this directory

Comment (hariszaf) [reply ext_data/community_models/e_coli_core.json]: We could remove the directory completely if you find it of no use.

File: ext_data/community_models/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+This is an example directory where you need to add all your `.mat` or `.json` model files 
+your community consists of. 
+
+For example, you may have a community of 2 models, *Escherichia coli* and *Helicobacter pylori*, 
+in that case you could get the [`e_coli_core.json`](http://bigg.ucsd.edu/models/e_coli_core) and 
+[`iIT341.json`](http://bigg.ucsd.edu/models/iIT341) files from the [BIGG database](http://bigg.ucsd.edu/models)
+and then run `dingo`. 
+
+Remember to remove this `README.md` file if you will decide to use this directory.
```
Comment (vissarion) [inline ext_data/community_models/README.md:9]: Why?

File: ext_data/community_models/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+This is an example directory where you need to add all your `.mat` or `.json` model files 
+your community consists of. 
+
+For example, you may have a community of 2 models, *Escherichia coli* and *Helicobacter pylori*, 
+in that case you could get the [`e_coli_core.json`](http://bigg.ucsd.edu/models/e_coli_core) and 
+[`iIT341.json`](http://bigg.ucsd.edu/models/iIT341) files from the [BIGG database](http://bigg.ucsd.edu/models)
+and then run `dingo`.
```
Comment (vissarion) [inline ext_data/community_models/README.md:7]: could you please add an example of use `i.e.` a command from terminal that uses dingo to do some computations to e_coli and iIT341 community

File: ext_data/community_models/README.md (hunk 2 above)
Comment (hariszaf) [reply ext_data/community_models/README.md:9]: If you leave the `README.md` file, `dingo` will give you an 
`Exception: An unknown input format given to initialize a metabolic network object.` message. 

We can add a check on the code for the `README` file if you feel this is necessary.

File: ext_data/community_models/README.md (hunk 3 above)
Comment (hariszaf) [reply ext_data/community_models/README.md:7]: Could you have a look now?

File: dingo/MetabolicNetwork.py [hunk 4]
```diff
@@ -223,3 +223,205 @@ def shut_down_reaction(self, index_val):
 
         self._lb[index_val] = 0
         self._ub[index_val] = 0
+
+
+class CommunityMetabolicNetwork():
```
Comment (TolisChal) [inline dingo/MetabolicNetwork.py]: Please use a single .py document per class

File: dingo/MetabolicNetwork.py [hunk 5]
```diff
@@ -223,3 +223,205 @@ def shut_down_reaction(self, index_val):
...
+        else:
+            return slow_fva(
+                self._comm_lb,
+                self._comm_ub,
+                self._comm_S,
+                self._comm_biomass_function,
+                self._comm_parameters["opt_percentage"],
+            )
+
+    def fba(self):
```
Comment (TolisChal) [inline dingo/MetabolicNetwork.py]: Which objective function do you use here?
I think FBA has to be separately applied in each model to obtain the maximum value per model.

File: dingo/MetabolicNetwork.py [hunk 6]
```diff
@@ -223,3 +223,205 @@ def shut_down_reaction(self, index_val):
...
+
+    @classmethod
+    def buildModelList(cls, directory, format_type):
+        comm_tuple_args = getModelList(directory, format_type)
+        return cls(comm_tuple_args)
+
+    def fva(self):
+        """A member function to apply the FVA method on the community metabolic network."""
+
+        if self._comm_parameters["fast_computations"]:
```
Comment (TolisChal) [inline dingo/MetabolicNetwork.py]: You should set the objective function to the zero vector before FVA() to study the unbiased case.

File: dingo/PolytopeSampler.py [hunk 7]
```diff
@@ -296,3 +295,228 @@ def set_tol(self, value):
     def set_opt_percentage(self, value):
 
         self._parameters["opt_percentage"] = value
+
+
+
+
+class CommunityPolytopeSampler:
```
Comment (TolisChal) [inline dingo/PolytopeSampler.py]: Please use a single .py document per class

File: dingo/MetabolicNetwork.py [hunk 8]
```diff
@@ -6,8 +6,8 @@
 # Licensed under GNU LGPL.3, see LICENCE file
 
 import numpy as np
-import sys
-from dingo.loading_models import read_json_file, read_mat_file
+import sys, os
+from dingo.loading_models import read_json_file, read_mat_file, getModelList
 from dingo.fva import slow_fva
 from dingo.fba import slow_fba
```
Comment (TolisChal) [inline dingo/MetabolicNetwork.py]: Please update the documentation according to your new implementations

File: dingo/MetabolicNetwork.py (hunk 4 above)
Comment (hariszaf) [reply dingo/MetabolicNetwork.py]: Done as suggested.

File: dingo/MetabolicNetwork.py (hunk 5 above)
Comment (hariszaf) [reply dingo/MetabolicNetwork.py]: We need to consider an FBA per organism (single standed network) and something like a community FBA for the total community model a.k.a the `μ_c` as described in the micom concept.. 

How about leaving as `fba` the per organism and build a new function for community growth ?

File: dingo/PolytopeSampler.py (hunk 7 above)
Comment (hariszaf) [reply dingo/PolytopeSampler.py]: Done as suggested.

**Comment by vissarion** [comment]
This is quite old and should be re-implemented from scratch

**Comment by hariszaf** [comment]
Way to go: 
- implement approach from [cobra toolbox](https://github.com/opencobra/cobratoolbox/blob/master/src/analysis/multiSpecies/microbiomeModelingToolbox/pairwiseInteractionModeling/createMultipleSpeciesModel.m) to build merged model 
- sample by asking biomass of species A to be 0 and biomass of species B to be max 
- sample with the vice versa conditions
- sample asking for both biomass functions to be max

**Comment by vissarion** [comment]
Thanks @hariszaf I added issue https://github.com/GeomScale/dingo/issues/67 to keep track on it!
